jython-3-roadmap.md

```diff
@@ -0,0 +1,293 @@
+---
+title: Jython 3 Roadmap
+---
+# Jython 3 Roadmap
+
+This discussion document attempts to outline the steps to Jython 3,
+defined by the MVP Features.
+There are probably glaring omissions.
+It is deliberately without dates.
+
+Apart from delivering the features,
+it aims to satisfy certain voluntary constraints,
+perceived as healthy in the long-term:
+
+* Comprehensible and documented architecture.
+  * Accessible for contribution.
+  * Readable as a model implementation of Python.
+* History is continuous with earlier contributed work, in spite of:
+  * Radical code re-organisation.
+  * Radical change (to some code).
+  * Rendering some code broken or dead (temporarily).
+
+In the interests of the second of these objectives (history),
+in the following,
+where it is implied we implement some class or package,
+the default approach will be to build on and credit prior art.
+To achieve this for each source file,
+we git-move, (commit) and afterwards modify
+the closest corresponding Jython 2 file.
+
+The middle commit will often produce a version that does not build,
+and should not be pushed to the project repository as the tip.
+Subsequent editing and another commit will correct that.
+Dead code,
+normally a Bad Thing,
+should remain until we know we won't resurrect it
+to supply a later feature.
+It may be necessary to create stubs to satisfy references in
+un-resurrected code.
+
+
+## A Sketchy Plan
+
+### Scorched Earth
+
+1. Restructure the code base so it builds with Gradle.
+Legacy code stays where it is,
+waiting to be moved and modified.
+
+1. The first build target is a library `jython-3.8a1-DEV`,
+but initially it will be empty.
+
+1. Provide some basic landmarks (modular project structure)
+and a convention for controlling log messages by sub-system.
+(It's a debugging aid for us
+and evolves to information for production use.)
+
+### Type and Arithmetic
+
+1. Outline architecture for objects, types, operations, slots.
+Specify the abstract object API (analogous CPython's).
+
+1. Implement `PyBaseObject`, `PyType`,
+and some simple types (mostly without operations).
+Implement only the Java API and write JUnit tests
+for type construction and inheritance.
+
+1. Implement type slots, add slot functions to simple types,
+and implement an abstract object API.
+Test the whole via JUnit, calling the abstract object API.
+From here on, new operations added imply
+new abstract object API and JUnit tests to match.
+
+1. Validation that acceptable performance is achieved,
+invoking arithmetic operations through the API.
+Likely to be measured using micro-benchmarks,
```

Comment:
Like the idea to add benchmarks, probably JMH, from as early as possible, it's these small core operations that will ultimately dominate real world performance, even if not much effort is directed to improve them at first, its good to have the framework there so when there is effort to focus on performance it can be well directed and measurable.

Reply:
I used JMH to check my work on some ` PyString` changes, and when I look at the old project, can't remember what it all means! But yes, JMH is what I meant. Useful as a canary in the coalmine.